apps/wappush/js/wappush.js

```diff
@@ -282,49 +266,127 @@
+  function wpm_onNew(obj) {
+    DUMP('A new message added to the DB');
+
+    var message = new ParsedMessage(obj);
+
+    if (message.action === 'signal-high' ||
+        message.action === 'execute-high')
+    {
+      /* We just decrease the number of pending messages instead of invoking
+       * wpm_finish() since we don't want to start the close procedure. */
+      pendingMessages--;
+      wpm_displayWapPushMessage(message.getUniqueId());
+    } else {
+      wpm_sendNotification(message).then(() => wpm_finish());
+    }
+  }
+
+  /**
+   * Reacts to a message being updated in the database by updating the
+   * notification and possibly the message being shown to the user.
+   *
+   * @param {Object} obj The new message in untyped form (i.e. JSON)
+   */
+  function wpm_onUpdate(obj) {
+    DUMP('A message was updated in the DB');
+
+    var message = new ParsedMessage(obj);
+
+    // Update the message if we're currently displaying it
+    if (message.getUniqueId() === displayedSiMessage.getUniqueId()) {
+      displayedSiMessage = message;
+      SiSlScreenHelper.populateScreen(message);
+    }
+
+    // This updates the existing notification
+    wpm_sendNotification(message).then(() => wpm_finish());
+  }
+
+  /**
+   * Removes the notifications for messages that have been deleted and closes
+   * the app if the message being viewed has been deleted.
+   *
+   * @param {Object} obj The deleted message in untyped form (i.e. JSON)
+   */
+  function wpm_onDelete(obj) {
+    DUMP('A message was deleted');
+
+    var message = new ParsedMessage(obj);
+
+    // Close the application if we're displaying the deleted message
+    if (displayedSiMessage &&
+      displayedSiMessage.getUniqueId() === message.getUniqueId()) {
+      wpm_close();
+    }
+
+    // Remove the notification associated to the deleted message
+    wpm_clearNotifications(+message.getUniqueId());
+  }
+
+  /**
+   * Prints a debug statement when a message has been discarded, no further
+   * action required.
+   */
+  function wpm_onDiscard() {
+    DUMP('The message was discarded');
+    wpm_finish();
   }
 
   /**
    * Retrieves a WAP Push message from the database and displays it
    *
-   * @param {String} timestamp The message timestamp as a string.
+   * @param {String} id The message id as a string.
    * @return {Promise} A promise that resolves once the message has been
    *                   displayed and the associated notification cleared.
    */
-  function wpm_displayWapPushMessage(timestamp) {
-    DUMP('Displaying message ' + timestamp);
+  function wpm_displayWapPushMessage(id) {
+    DUMP('Displaying message ' + id);
 
     // Clear the close timer as the application will soon become visible
     app.launch();
     window.clearTimeout(closeTimeout);
     closeTimeout = null;
 
-    return ParsedMessage.load(timestamp).then(
+    return ParsedMessage.load(id).then(
       function wpm_loadResolved(message) {
-        if (message) {
```

Comment:
The real reason why I removed it is that we can't have notifications for deleted messages anymore so it shouldn't be possible to call MessageDB.retrieve() on a deleted message anymore. I'll make mdb_retrieve() throw if the message is not found so that it will be harder to misuse it in the future by trying to retrieve messages that don't exist.